Ignore endpoint events if processed already.

aswinsuryan opened the pull request:
In deployments, if we delete submariner-gateway machineset and then
later delete the submariner namespace in a cluster. It would lead
to a stale endpoint in other clusters. Now if we create a machineset
and join the cluster again, the old endpoint gets cleared. But the old
endpoint delete event comes after the new endpoint is created and the
host network routing rules gets deleted.

With this fix, a delete event will be ignored if it has timestamp older
than the current endpoint.

Fixes: #1544

Signed-off-by: Aswin Surayanarayanan <asuryana@redhat.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our developer guide: https://submariner.io/development/
2. Ensure you have added the appropriate tests for your PR: https://submariner.io/development/code-review/#test-new-functionality
3. Read the code review guide to ease the review process: https://submariner.io/development/code-review/
4. If the PR is unfinished, mark it as a draft: https://submariner.io/development/code-review/#mark-work-in-progress-prs-as-drafts
5. If you are using CI to debug, use your private fork: https://submariner.io/development/code-review/#use-private-forks-for-debugging-prs-by-running-ci
6. Add labels to the PR as appropriate.

This template is based on the K8s/K8s template:

https://github.com/kubernetes/kubernetes/blob/master/.github/PULL_REQUEST_TEMPLATE.md
-->

tpantelis commented on pkg/routeagent_driver/handlers/kubeproxy/endpoint_handler.go:
```diff
@@ -101,6 +101,15 @@ func (kp *SyncHandler) RemoteEndpointCreated(endpoint *submV1.Endpoint) error {
 	kp.syncHandlerMutex.Lock()
 	defer kp.syncHandlerMutex.Unlock()
+
+	lastPorcessedTime, ok := kp.remoteEndpointTimeStamp[endpoint.Name]
```
```suggestion
	lastProcessedTime, ok := kp.remoteEndpointTimeStamp[endpoint.Name]
```

tpantelis commented on pkg/routeagent_driver/handlers/kubeproxy/endpoint_handler.go:
```diff
@@ -101,6 +101,15 @@ func (kp *SyncHandler) RemoteEndpointCreated(endpoint *submV1.Endpoint) error {
+	lastPorcessedTime, ok := kp.remoteEndpointTimeStamp[endpoint.Name]
+
+	if ok && lastPorcessedTime.After(endpoint.CreationTimestamp.Time) {
+		klog.Infof("Ingoring new remote %#v since a later endoint was already"+
```
I'm not sure how noisy this could be but It seems this should be DEBUG or TRACE level.

sridhargaddam replied:
Ideally, these events should happen only during migration and if there is out-of-order events. So, I would guess it should be less noisy.

aswinsuryan replied:
Ya it should happen only in corner conditions and should be good for us to know if it happens.

tpantelis commented on pkg/routeagent_driver/handlers/kubeproxy/endpoint_handler.go:
```diff
@@ -129,6 +140,17 @@ func (kp *SyncHandler) RemoteEndpointUpdated(endpoint *submV1.Endpoint) error {
 	kp.syncHandlerMutex.Lock()
 	defer kp.syncHandlerMutex.Unlock()
+
+	lastPorcessedTime, ok := kp.remoteEndpointTimeStamp[endpoint.Name]
```
```suggestion
	lastProcessedTime, ok := kp.remoteEndpointTimeStamp[endpoint.Name]
```

aswinsuryan replied:
Done.

tpantelis commented on pkg/routeagent_driver/handlers/kubeproxy/endpoint_handler.go:
```diff
@@ -129,6 +140,17 @@ func (kp *SyncHandler) RemoteEndpointUpdated(endpoint *submV1.Endpoint) error {
+	lastPorcessedTime, ok := kp.remoteEndpointTimeStamp[endpoint.Name]
+
+	if ok && lastPorcessedTime.After(endpoint.CreationTimestamp.Time) {
+		klog.Infof("Ingoring deleted remote %#v since a later endoint was already"+
```
DEBUG or TRACE?

aswinsuryan replied:
Retaining Info as mentioned above.

tpantelis reviewed (commented):
Can you explain in the PR summary the proposed solution and how it fixes the issue?

sridhargaddam commented on pkg/routeagent_driver/handlers/kubeproxy/endpoint_handler.go:
```diff
@@ -129,6 +140,17 @@ func (kp *SyncHandler) RemoteEndpointUpdated(endpoint *submV1.Endpoint) error {
+	if ok && lastPorcessedTime.After(endpoint.CreationTimestamp.Time) {
+		klog.Infof("Ingoring deleted remote %#v since a later endoint was already"+
+			"processed", endpoint)
+		return nil
```
The exact problem that we noticed in OCP deployment was that a stale/old endpoint event coming after an updated event. Here we are skipping the old one, but what about deleting any resources (like iptables, routes etc) which point to the old endpoint? Can you please point me where we will handle it with this change?

tpantelis replied:
This code correlates the time stamp between endpoints with the same name. I'm curious what sequence of events can occur for endpoint(s) with the same name that would cause an issue? Can the exact scenario sequence that we're trying to solve be outlined in the PR?

aswinsuryan replied:
Added more details to the commit message.

sridhargaddam replied:
@aswinsuryan, thanks for the PR. I see that you are now skipping the endpoint processing if its arriving after the new endpoint. Can you please check that on the old gateway node, the following entries are removed. We are seeing this more often in OCP setup.

1. IP route table 150 entries
2. vx-submariner interface 
3. IP xfrm policy/state rules

aswinsuryan replied:
@sridhargaddam This issue occurs when the old g/w node(machine set) is deleted and a new one is created. So in the remote cluster, the old one does not exist. In the local cluster, g/w node did not change, but it received a stale event. So we need to keep that rules.

Is  that what you meant to do?

dfarrell07 replied:
> This issue occurs when the old g/w node(machine set) is deleted and a new one is created.

Is it possible to add tests for this scenario? I'm thinking of the [CII Silver](https://bestpractices.coreinfrastructure.org/en/criteria) requirement: 

> The project MUST add regression tests to an automated test suite for at least 50% of the bugs fixed within the last six months.

aswinsuryan replied:
@dfarrell07 I am not sure if it can be done in kind,may be removing a worker node and adding it back help.

sridhargaddam reviewed (commented):
Thanks for the PR @aswinsuryan. I have one comment.

sridhargaddam commented:
@aswinsuryan ping.

aswinsuryan committed:
Fix issue with gateway name in e2e

The submariner gateway name is not always the same as the node name.
It is generated using an API, hence using the same API for test
as well, to avoid failures in some scenarios.

Signed-off-by: Aswin Suryanarayanan <aswinsuryan@gmail.com>

aswinsuryan commented:
> @aswinsuryan ping.
Replied in the comments.

sridhargaddam commented on pkg/routeagent_driver/handlers/kubeproxy/endpoint_handler.go:
```diff
@@ -135,6 +145,16 @@ func (kp *SyncHandler) RemoteEndpointRemoved(endpoint *submV1.Endpoint) error {
+	lastProcessedTime, ok := kp.remoteEndpointTimeStamp[endpoint.ClusterName]
+
+	if ok && lastProcessedTime.After(endpoint.CreationTimestamp.Time) {
+		klog.Infof("Ignoring deleted remote %#v since a later endoint was already"+
+			"processed", endpoint)
```
```suggestion
		klog.Infof("Ignoring deleted remote %#v since a later endpoint was already"+
			"processed", endpoint)
```

sridhargaddam commented on pkg/routeagent_driver/handlers/kubeproxy/endpoint_handler.go:
```diff
@@ -105,6 +105,14 @@ func (kp *SyncHandler) RemoteEndpointCreated(endpoint *submV1.Endpoint) error {
+	lastProcessedTime, ok := kp.remoteEndpointTimeStamp[endpoint.ClusterName]
+
+	if ok && lastProcessedTime.After(endpoint.CreationTimestamp.Time) {
+		klog.Infof("Ignoring new remote %#v since a later endoint was already"+
```
```suggestion
		klog.Infof("Ignoring new remote %#v since a later endpoint was already"+
```

sridhargaddam reviewed (commented):
Thanks @aswinsuryan. Some minor nits
